File: typescript/GameOfLife.TypeScript.Web/src/services/game-service.ts
import { HubConnection, HubConnectionBuilder } from "@aspnet/signalr";
import { World } from "../models/World";

export class GameService {
    private connection: HubConnection;

    constructor() {
        console.log(process.env.REACT_APP_API_URL);
        const websocketUrl = `${process.env.REACT_APP_API_URL}/game`;
        this.connection = new HubConnectionBuilder()
            .withUrl(websocketUrl)
            .build();
    }
    
    public connect() {
        return this.connection.start();
    }

    public subscribe(func: (world: World) => void) {
        this.connection.on("UpdateGameWorld", func);
    }
    
    public start(userId: number, patternId: number) {
        return this.connection.invoke("StartGameFromPattern", userId, patternId)
    }

    public end(userId: number) {
        return this.connection.invoke("EndUserGame", userId)
    }
}

File: csharp/GameOfLife.CSharp.Web/Data/GameService.cs
using GameOfLife.CSharp.Web.Models;
using Microsoft.AspNetCore.SignalR.Client;
using System;
using System.Threading.Tasks;

namespace GameOfLife.CSharp.Web.Data
{
    public class GameService : IGameService
    {
        private readonly HubConnection _hubConnection;

        public GameService()
        {
            _hubConnection = new HubConnectionBuilder()
             .WithUrl("https://localhost:44370/game")
             .Build();
        }

        public async Task Connect()
        {
            if (_hubConnection.State == HubConnectionState.Disconnected)
            {
                await _hubConnection.StartAsync();
            }
        }

        public async Task End(int userId)
        {
            await _hubConnection.SendAsync("EndUserGame", userId);
        }

        public async Task Start(int userId, int patternId)
        {
            await _hubConnection.SendAsync("StartGameFromPattern", userId, patternId);
        }

        public void Subscribe(Action<World> action)
        {
            _hubConnection.On("UpdateGameWorld", action);
        }
    }
}
